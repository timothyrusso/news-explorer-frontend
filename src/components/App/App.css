.content {
  max-width: 1440px;
  margin: auto;
  font-family: 'Inter', Arial, sans-serif;
  border: 2px solid red;
  width: 100%;
  min-height: 100vh;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-text-size-adjust: 100%;
  text-rendering: optimizeLegibility;
}
