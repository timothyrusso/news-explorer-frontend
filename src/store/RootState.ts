import { UserState } from './user/user.reducer';
import { ArticleState } from './article/article.reducer';
import { TogglesState } from './toggles/toggles.reducer';
import { ErrorsState } from './errors/errors.reducer';

export type RootState = {
  user: UserState;
  article: ArticleState;
  toggles: TogglesState;
  errors: ErrorsState;
};
