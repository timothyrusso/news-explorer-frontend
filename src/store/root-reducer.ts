import { combineReducers } from 'redux';
import { userReducer } from './user/user.reducer';
import { articleReducer } from './article/article.reducer';

export const rootReducer = combineReducers({
  user: userReducer,
  article: articleReducer,
});
