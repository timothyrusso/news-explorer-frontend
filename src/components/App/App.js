import "./App.css";
import Navigation from "../Navigation/Navigation";
import Header from "../Header/Header";
import Main from "../Main/Main";
import Footer from "../Footer/Footer";
import SigninPopup from "../SigninPopup/SigninPopup";
import SignupPopup from "../SignupPopup/SignupPopup";
import NothingFound from "../NothingFound/NothingFound";
import { useState, useEffect, useLayoutEffect } from "react";
import InfoTooltip from "../InfoTooltip/InfoTooltip";
import Preloader from "../Preloader/Preloader";
import NewsCardList from "../NewsCardList/NewsCardList";
import { cards } from "../../utils/testCards";
import {
  Route,
  Routes,
  useLocation,
  Navigate,
  useNavigate,
} from "react-router-dom";
import SavedNewsHeader from "../SavedNewsHeader/SavedNewsHeader";
import SavedNews from "../SavedNews/SavedNews";
import { getNewsInfo } from "../../utils/api";

const App = () => {
  const [isSigninPopupOpen, setIsSigninPopupOpen] = useState(false);
  const [isSignupPopupOpen, setIsSignupPopupOpen] = useState(false);
  const [isInfoTooltipOpen, setIsInfoTooltipOpen] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [formValidity, setFormValidity] = useState(true);
  const [errorMessage, setErrorMessage] = useState({});
  const [popupRedirectText, setPopupRedirectText] = useState("");
  const [loggedIn, setLoggedIn] = useState(true);
  const [blackNavigator, setBlackNavigator] = useState(false);
  const [savedCard, setSavedCard] = useState(false);
  const [showNews, setShowNews] = useState(false);
  const [totalResults, setTotalResults] = useState(1);
  const [page, setPage] = useState(1);
  const [search, setSearch] = useState("");
  const [toggleMenu, setToggleMenu] = useState(false);
  const [newsArticles, setNewsArticles] = useState([]);

  const location = useLocation();
  const history = useNavigate();

  const resultsPerPage = 3;
  const showMoreLogic = page * resultsPerPage < totalResults;

  const closeAllPopups = () => {
    setIsSigninPopupOpen(false);
    setIsSignupPopupOpen(false);
    setIsInfoTooltipOpen(false);
  };

  const startLoading = () => {
    setIsLoading(true);
  };

  const startLoadingNews = () => {
    setShowNews(false);
    setIsLoading(true);
  };

  const onFormUpdate = (data) => {
    data ? setFormValidity(true) : setFormValidity(false);
  };

  const checkValidity = (evt) => {
    const name = evt.target.name;
    setErrorMessage({ ...errorMessage, [name]: evt.target.validationMessage });
  };

  const handleSigninPopupClick = () => {
    setFormValidity(true);
    setErrorMessage({});
    closeAllPopups();
    setPopupRedirectText("Sign up");
    setIsSigninPopupOpen(true);
  };

  const handleSwitchPopup = () => {
    // da sistemare e rendere piÃ¹ elegante
    if (isSigninPopupOpen) {
      setIsSigninPopupOpen(false);
      setIsSignupPopupOpen(true);
      setPopupRedirectText("Sign in");
    } else if (isSignupPopupOpen) {
      setIsSignupPopupOpen(false);
      setIsSigninPopupOpen(true);
      setPopupRedirectText("Sign up");
    }
  };

  const handleLogout = () => {
    setLoggedIn(false);
    history("/");
  };

  const setNewPage = () => {
    setPage(page + 1);
    showMoreResults();
  };

  const toggleNav = () => {
    setToggleMenu(!toggleMenu);
  };

  const showMoreResults = () => {
    getNewsInfo({ search }, page + 1)
      .then((data) => {
        setTotalResults(data.totalResults);
        setNewsArticles([...newsArticles, ...data.articles]);
      })
      .finally(() => {
        setIsLoading(false);
        setShowNews(true);
      });
  };

  const activateSearch = (data) => {
    getNewsInfo({ search: data.search }, page)
      .then((data) => {
        setTotalResults(data.totalResults);
        setNewsArticles(data.articles);
      })
      .finally(() => {
        setIsLoading(false);
        setShowNews(true);
        setSearch(data.search);
      });
  };

  useEffect(() => {
    switch (location.pathname) {
      case "/":
        setBlackNavigator(false);
        setSavedCard(false);
        setToggleMenu(false);
        return;
      case "/saved-news":
        setBlackNavigator(true);
        setSavedCard(true);
        setToggleMenu(false);
        return;
    }
  }, [location, setBlackNavigator, setSavedCard, setToggleMenu]);

  useLayoutEffect(() => {
    window.scrollTo(0, 0);
  }, [location]);

  return (
    <div className="content">
      <Navigation
        onSigninPopupClick={handleSigninPopupClick}
        loggedIn={loggedIn}
        blackNavigator={blackNavigator}
        handleLogout={handleLogout}
        toggleMenu={toggleMenu}
        toggleNav={toggleNav}
      />
      <Routes>
        <Route
          exact
          path="/"
          element={
            <>
              <Header
                startLoadingNews={startLoadingNews}
                activateSearch={activateSearch}
              />
              {showNews && newsArticles.length !== 0 && (
                <NewsCardList
                  cards={cards}
                  onSigninPopupClick={handleSigninPopupClick}
                  loggedIn={loggedIn}
                  newsArticles={newsArticles}
                  showMoreResults={showMoreResults}
                  setNewPage={setNewPage}
                  showMoreLogic={showMoreLogic}
                />
              )}
              {newsArticles.length === 0 && !isLoading && <NothingFound />}
              {isLoading && <Preloader />}
              <Main />
              <SigninPopup
                isOpen={isSigninPopupOpen}
                onClose={closeAllPopups}
                onSwitch={handleSwitchPopup}
                popupRedirectText={popupRedirectText}
                isLoading={isLoading}
                startLoading={startLoading}
                formValidity={formValidity}
                onFormUpdate={onFormUpdate}
                errorMessage={errorMessage}
                onInputUpdate={checkValidity}
              />
              <SignupPopup
                isOpen={isSignupPopupOpen}
                onClose={closeAllPopups}
                onSwitch={handleSwitchPopup}
                popupRedirectText={popupRedirectText}
                isLoading={isLoading}
                startLoading={startLoading}
                formValidity={formValidity}
                onFormUpdate={onFormUpdate}
                errorMessage={errorMessage}
                onInputUpdate={checkValidity}
              />
              <InfoTooltip
                isOpen={isInfoTooltipOpen}
                onClose={closeAllPopups}
                openSignin={handleSigninPopupClick}
              />
            </>
          }
        />
        <Route
          path="/saved-news"
          element={
            <>
              <SavedNewsHeader />
              <SavedNews cards={cards} savedCard={savedCard} />
            </>
          }
        />
        <Route path="*" element={<Navigate to="/" replace />} />
      </Routes>
      <Footer />
    </div>
  );
};

export default App;
