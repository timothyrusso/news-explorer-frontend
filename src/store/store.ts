import { legacy_createStore as createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { rootReducer } from './root-reducer';
import { ThunkDispatch } from 'redux-thunk';
import { AnyAction } from 'redux';
import { RootState } from './RootState';

const store = createStore(rootReducer, applyMiddleware(thunk));

export type AppDispatch = ThunkDispatch<RootState, unknown, AnyAction>;

export default store;
