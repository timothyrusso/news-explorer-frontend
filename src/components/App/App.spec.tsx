import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from '../../store/store';
import { render } from '@testing-library/react';
import App from './App';

describe('App component', () => {
  it('Should render the App component', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Router>
          <App />
        </Router>
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
