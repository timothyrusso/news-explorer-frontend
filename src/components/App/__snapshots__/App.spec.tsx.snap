// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component Should render the App component 1`] = `
<DocumentFragment>
  <div
    class="content"
  >
    <section
      class="navigation "
    >
      <nav
        class="navigation__bar "
      >
        <a
          class="logo"
          href="/"
        >
          <img
            alt="NewsExplorer logo"
            src="test-file-stub"
          />
        </a>
        <ul
          class="navigation__links-wrapper"
        >
          <a
            class="navigation__link               navigation__link_type_focus"
            href="/"
            style="color: white; border-color: white;"
          >
            Home
          </a>
          <a
            class="navigation__link navigation__link_type_saved navigation__disabled
              
              "
            href="/saved-news"
            style="color: white; border-color: white;"
          >
            Saved articles
          </a>
          <button
            aria-label="signin"
            class="navigation__button navigation__button_type_signin "
            style="color: white; border-color: white;"
            type="button"
          >
            Sign in
          </button>
          <button
            aria-label="signout"
            class="navigation__button navigation__button_type_logout navigation__disabled"
            style="color: white; border-color: white;"
            type="button"
          >
            <img
              alt="Logout logo"
              class="navigation__logout-icon"
              src="test-file-stub"
            />
          </button>
        </ul>
        <button
          class="burgher-button
          false
          burgher-button_type_white"
        />
      </nav>
    </section>
    <header
      class="header"
    >
      <h1
        class="header__title"
      >
        What's going on in the world?
      </h1>
      <p
        class="header__subtitle"
      >
        Find the latest news on any topic and save them in your personal account.
      </p>
      <form
        class="search-form"
      >
        <input
          class="search-form__input "
          id="search-input"
          name="search"
          placeholder="Enter topic"
          type="text"
          value=""
        />
        <button
          class="search-form__button"
        >
          Search
        </button>
      </form>
    </header>
    <main
      class="main"
    >
      <section
        class="about"
      >
        <img
          alt="Timothy - a passionate software engineer"
          class="about__image"
          src="test-file-stub"
        />
        <article
          class="about__wrapper"
        >
          <h2
            class="about__title"
          >
            About the author
          </h2>
          <p
            class="about__description"
          >
            Hey there, I'm Timothy. I'm a motivated and passionate software engineer with a background in eCommerce, Quality Assurance, Application Delivery Management, Project Management and Trade Marketing. I have a bachelor's degree in Economics and Business Management and I have done a coding bootcamp called TripleTen.
          </p>
          <p
            class="about__description"
          >
            I’m a curious person, I love to code and I like to experiment with new technologies. Now coding is my passion and my professional career focus.
          </p>
        </article>
      </section>
    </main>
    <div
      class="popup  popup_type_signin"
    >
      <div
        class="popup__container"
      >
        <form
          action="#"
          class="popup__form"
          name="myFormmyFormsignin"
        >
          <h2
            class="popup__title"
          >
            Sign in
          </h2>
          <label
            class="popup__input-label"
            for="email-signin-input"
          >
            Email
          </label>
          <input
            class="popup__input popup__input_field_email"
            id="email-signin-input"
            name="email"
            placeholder="Enter email"
            required=""
            type="email"
            value=""
          />
          <span
            class="popup__input_type_error"
            id="email-signin-input-error"
          />
          <label
            class="popup__input-label"
            for="password-signin-input"
          >
            Password
          </label>
          <input
            class="popup__input popup__input_field_password"
            id="password-signin-input"
            maxlength="20"
            minlength="8"
            name="password"
            placeholder="Enter password"
            required=""
            type="password"
            value=""
          />
          <span
            class="popup__input_type_error"
            id="password-signin-input-error"
          />
          <p
            aria-live="polite"
            class="popup-server-error"
          />
          <button
            aria-label="Submit Sign in"
            class="submit-button popup__button  "
            type="submit"
          >
            Sign in
          </button>
          <p
            class="form-redirect"
          >
            or 
            <button
              aria-label="test"
              class="form-redirect_type_link"
              type="button"
            />
          </p>
        </form>
        <button
          aria-label="Close signin popup"
          class="close-button "
          type="button"
        />
      </div>
    </div>
    <div
      class="popup  popup_type_signup"
    >
      <div
        class="popup__container"
      >
        <form
          action="#"
          class="popup__form"
          name="myFormmyFormsignup"
        >
          <h2
            class="popup__title"
          >
            Sign up
          </h2>
          <label
            class="popup__input-label"
            for="email-input"
          >
            Email
          </label>
          <input
            class="popup__input popup__input_field_email"
            id="email-input"
            maxlength="254"
            minlength="5"
            name="email"
            placeholder="Enter email"
            required=""
            type="email"
            value=""
          />
          <span
            class="popup__input_type_error"
            id="email-input-error"
          />
          <label
            class="popup__input-label"
            for="password-input"
          >
            Password
          </label>
          <input
            class="popup__input popup__input_field_password"
            id="password-input"
            maxlength="20"
            minlength="8"
            name="password"
            placeholder="Enter password"
            required=""
            type="password"
            value=""
          />
          <span
            class="popup__input_type_error"
            id="password-input-error"
          />
          <label
            class="popup__input-label"
            for="username-input"
          >
            Username
          </label>
          <input
            class="popup__input popup__input_field_username"
            id="username-input"
            maxlength="20"
            minlength="2"
            name="username"
            placeholder="Enter your username"
            required=""
            type="text"
            value=""
          />
          <span
            class="popup__input_type_error"
            id="username-input-error"
          />
          <p
            aria-live="polite"
            class="popup-server-error"
          />
          <button
            aria-label="Submit Sign up"
            class="submit-button popup__button  "
            type="submit"
          >
            Sign up
          </button>
          <p
            class="form-redirect"
          >
            or 
            <button
              aria-label="test"
              class="form-redirect_type_link"
              type="button"
            />
          </p>
        </form>
        <button
          aria-label="Close signup popup"
          class="close-button "
          type="button"
        />
      </div>
    </div>
    <div
      class="popup  popup_type_undefined"
    >
      <div
        class="popup__container"
      >
        <div
          class="info-tooltip"
        >
          <p
            class="info-tooltip__text"
          >
            Registration successfully completed!
          </p>
          <button
            aria-label="signin"
            class="info-tooltip__button"
            type="button"
          >
            Sign in
          </button>
        </div>
        <button
          aria-label="Close undefined popup"
          class="close-button close-button_type_info"
          type="button"
        />
      </div>
    </div>
    <footer
      class="footer"
    >
      <p
        class="footer__copyright"
      >
        © 2023 NewsExplorer - Powered by 
        <a
          href="https://newsapi.org/"
          rel="noreferrer"
          target="_blank"
        >
          News API
        </a>
      </p>
      <nav
        class="footer__wrapper"
      >
        <a
          class="footer__link"
          href="/"
        >
          Home
        </a>
        <a
          class="footer__link footer__link_type_mobile"
          href="https://timothyrusso.dev/"
          rel="noreferrer"
          target="_blank"
        >
          Portfolio
        </a>
        <ul
          class="footer__icon-wrapper"
        >
          <a
            href="https://github.com/timothyrusso"
            rel="noreferrer"
            target="_blank"
          >
            <img
              alt="GitHub logo"
              class="footer__icon"
              src="test-file-stub"
            />
          </a>
          <a
            href="https://www.linkedin.com/in/russotimothysoftwareengineer/"
            rel="noreferrer"
            target="_blank"
          >
            <img
              alt="LinkedIn Logo"
              class="footer__icon"
              src="test-file-stub"
            />
          </a>
        </ul>
      </nav>
    </footer>
  </div>
</DocumentFragment>
`;
